### ISSUE TITLE
Model not learning anything during training

### ISSUE THREAD

**Issue by arthursw** [MedICL-VU/LesionSeg#9]
I started a training with the following command:

`python train.py --loss_to_use focal --num_threads 0 --input_nc 3 --batch_size 16 --trainSize 300 --testSize 300 --test_mode val --name emisep --eval_val`

But after 130 epochs the model didn't learn anything, it seems it does not predict any lesion:

<img width="1850" alt="LesionSegTraining" src="https://user-images.githubusercontent.com/1814799/90889066-b2697f00-e3b7-11ea-8fdb-5065c60c1dda.png">

There are 24 patient with flair, t1 and t2 modalities in my training set.

Here is my configurations.py
```
import os

PATH_DATASET = os.path.join('/home', 'amasson', 'ms_study', 'methods', 'lesion_seg', 'data', 'input', 'training')

MODALITIES = ['flair', 't1', 't2']
MASKS = ['mask']
SUFFIX = 'nii.gz'

AXIS_TO_TAKE = [2, 0, 1]

```

By the way, I don't know how to find the axis order of my data (to set `AXIS_TO_TAKE`) so this setting might be wrong, but I guess it is not the only reason why the training does not work.

Any idea?

Here are a few logs concerning my settings:

```
(base) amasson@niseko:~/ms_study/methods/lesion_seg/data/input/training$ cat ids.json 
{
  "59": {
    "1": {
      "modalities": {
        "flair": "/home/amasson/ms_study/methods/lesion_seg/data/input/training/raw/montpellier-20170112_59_1_flair.nii.gz",
        "t1": "/home/amasson/ms_study/methods/lesion_seg/data/input/training/raw/montpellier-20170112_59_1_t1.nii.gz",
        "t2": "/home/amasson/ms_study/methods/lesion_seg/data/input/training/raw/montpellier-20170112_59_1_t2.nii.gz"
      },
      "mask": {
        "mask": "/home/amasson/ms_study/methods/lesion_seg/data/input/training/raw/montpellier-20170112_59_1_mask.nii.gz"
      }
    },
    "2": { ... }
  },
  ...
}

(base) amasson@niseko:~/ms_study/methods/lesion_seg/data/input/training$ cat properties.json 
{
  "50ab42594bb6f9e5e4b73b4601b32324e6000243": {
    "path": "/home/amasson/ms_study/methods/lesion_seg/data/input/training/raw/montpellier-20170112_59_1_flair.nii.gz",
    "peak": 105.548234933748
  },
  ...
}

(base) amasson@niseko:~/ms_study/methods/lesion_seg/data/input/training$ tree .
.
├── ids.json
├── properties.json
├── raw
│   ├── montpellier-20170112_59_1_flair.nii.gz
│   ├── montpellier-20170112_59_1_mask.nii.gz
│   ├── montpellier-20170112_59_1_t1.nii.gz
│   ├── montpellier-20170112_59_1_t2.nii.gz
│   ├── montpellier-20170112_59_2_flair.nii.gz
│   ├── montpellier-20170112_59_2_mask.nii.gz
│   ├── montpellier-20170112_59_2_t1.nii.gz
│   ├── montpellier-20170112_59_2_t2.nii.gz
...
├── test
├── train
│   ├── 0.pkl
│   ├── 1000.pkl
...
└── val
    ├── mask_val10_flair.nii.gz
    ├── mask_val10_mask.nii.gz
    ├── mask_val10_t1.nii.gz
    ├── mask_val10_t2.nii.gz
    ...

4 directories, 9261 files

(base) amasson@niseko:~/ms_study/methods/lesion_seg/data/input/training$ space .
2,3G ./raw
492M ./val
4,0K ./test
15G	 ./train
18G	 .
```

**Comment by arthursw** [comment]
It looks better with L2 loss (I kept the same train & test size of 300):

![lesionSegL2Loss](https://user-images.githubusercontent.com/1814799/91441744-0caf8780-e871-11ea-9e66-7466714ffb28.png)

**Comment by arthursw** [comment]
yet I wonder why there is so much empty space around the brain, why it is not cropped to only contain the brain and minimize the number of zeros? Is this just an effect of the visualization?

**Comment by arthursw** [comment]
Maybe I should resample my images so that I get something more homogeneous in all directions.

**Comment by HuahongZhang** [comment]
> yet I wonder why there is so much empty space around the brain, why it is not cropped to only contain the brain and minimize the number of zeros? Is this just an effect of the visualization?
> Or maybe the program adds padding to precisely reach 300 voxels in each direction whereas my images are approximately 300 x 285 x 126 ; so going 126 (the smallest dimension) to 300 would create a lot of unused zeros.

Hi Arthur, if you want to make it cropped to only contain the brain, you can use smaller trainSize but it may not contains the whole brain. As I think, it is OK to use a trainSize like 128. And you are right, resampling the images would be a great idea.

**Comment by arthursw** [comment]
Ok, my results are looking good 👍  now I'll try to resample the brain to get better performances. 
I assume the bigger image the better, right? 
And I also guess it's better to have isotropic sampling (same spacing in all directions) ?

Meaning I will resample my image, crop them tightly (loosing a tiny bit or the borders), and to get 300 voxels wide volumes.

Or maybe dense an-isotropic data (without zero-padding) is better than isotropic data with some zero-padding?

**Comment by HuahongZhang** [comment]
As I think, a trainSize slightly smaller than the size of any direction (for isotropic image) performs well, but a larger trainSize maybe better. About the need to do resampling, I am not pretty sure since I have not compared the performance before and after resampling for anisotropic data, but I think it would be easy for you to implement and test :)

**Comment by arthursw** [comment]
Yes, thanks a lot for your help :) !
